### PR TITLE
Add Phase 2 & 3 training to prototype

### LINKED ISSUE #33: Add phase 2 and phase 3 training to prototype

### PR DESCRIPTION
**Pull request by DrKwint**
A fully implemented training routine for Prototype. Closes #33, but if this merges, open a new issue for debugging the prototype push procedure which is reducing accuracy to guessing (phase 2 of training)

### CODE REVIEW

File: pyroclast/prototype/model.py [hunk 1]
```diff
@@ -60,20 +67,38 @@ def __init__(self,
         else:
             self.classifier = tf.keras.layers.Dense(num_classes, use_bias=False)
 
+    @property
+    def trainable_prototype_vars(self):
+        return list(self.prototype_layer.trainable_variables)
+
+    @property
+    def final_conv_vars(self):
+        self.final_conv.trainable_variables
```
Comment (ihowell) [inline pyroclast/prototype/model.py]: You forgot to return the value?

File: pyroclast/prototype/prototype.py [hunk 2]
```diff
@@ -96,17 +113,101 @@ def run_minibatch(epoch, batch, is_train=True):
...
+                                  step=global_step)
             tf.summary.scalar(prefix + "loss/total loss",
                               loss,
                               step=global_step)
 
+    ### PHASE 1
     # run training loop
-    for epoch in range(epochs):
+    if debug: print("PHASE 1 - TRAINING CONV STACK AND PROTOTYPES")
+    for epoch in range(epochs_phase_1):
```
Comment (ihowell) [inline pyroclast/prototype/prototype.py]: Shouldn't this block of code only run if phase is equal to one?

File: pyroclast/prototype/prototype.py [hunk 3]
```diff
@@ -96,17 +113,101 @@ def run_minibatch(epoch, batch, is_train=True):
...
+        # test
+        test_batches = data_dict['test']
+        if debug:
+            print("TEST")
+            test_batches = tqdm(test_batches, total=data_dict['test_bpe'])
+        for batch in test_batches:
+            run_minibatch(epoch, batch, phase=1, is_train=False)
+
+    ### PHASE 2
+    if debug: print("PHASE 2 - PUSHING PROTOTYPES")
```
Comment (ihowell) [inline pyroclast/prototype/prototype.py]: And this code only run if phase == 2?

File: pyroclast/prototype/prototype.py [hunk 4]
```diff
@@ -96,17 +113,101 @@ def run_minibatch(epoch, batch, is_train=True):
...
+            [-1])
+        # set new value
+        new_prototypes[i] = np.reshape(class_patches[label],
+                                       [-1])[np.argmin(distances)]
+    model.prototypes = new_prototypes
+    print("Classification rate after prototype push: ",
+          classification_rate(data_dict['train']))
+
+    ### PHASE 3
+    if debug: print("PHASE 3 - TRAINING CLASSIFIER")
```
Comment (ihowell) [inline pyroclast/prototype/prototype.py]: And this 3?

**Comment by DrKwint** [comment]
[Something](https://trello.com/c/aMkUAJlr/1-something)

File: pyroclast/prototype/prototype.py (hunk 2 above)
Comment (DrKwint) [reply pyroclast/prototype/prototype.py]: It's the part of the code that actually runs the first phase. Ditto for the following 2 sections as well.

File: pyroclast/prototype/model.py (hunk 1 above)
Comment (DrKwint) [reply pyroclast/prototype/model.py]: Thanks, fixed in next commit!

**Comment by DrKwint** [comment]
Requesting final approval from @ihowell

File: pyroclast/prototype/prototype.py (hunk 2 above)
Comment (ihowell) [reply pyroclast/prototype/prototype.py]: That's what I'm saying. The first phase should be run if the phase argument is set to 1, right? The argument doesn't seem to actually do what it's intended to do.
